src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/StorePal.LoaderPal.cs

```diff
@@ -56,5 +56,74 @@ public void MoveTo(X509Certificate2Collection collection)
                 }
             }
         }
+
+        private sealed class ApplePkcs12CertLoader : ILoaderPal
+        {
+            private readonly ApplePkcs12Reader _pkcs12;
+            private readonly SafeKeychainHandle _keychain;
+            private readonly SafePasswordHandle _password;
+            private readonly bool _exportable;
+
+            public ApplePkcs12CertLoader(
+                ApplePkcs12Reader pkcs12,
+                SafeKeychainHandle keychain,
+                SafePasswordHandle password,
+                bool exportable)
+            {
+                _pkcs12 = pkcs12;
+                _keychain = keychain;
+                _exportable = exportable;
+
+                bool addedRef = false;
+                password.DangerousAddRef(ref addedRef);
+                _password = password;
+            }
+
+            public void Dispose()
+            {
+                _pkcs12.Dispose();
+
+                // Only dispose the keychain if it's a temporary handle.
+                (_keychain as SafeTemporaryKeychainHandle)?.Dispose();
+
+                _password.DangerousRelease();
```

Comment:
this looks like potential double free issue, should you guard against it or use regular Dispose?